pkg/engine/plugins.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package engine
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pulumi/pulumi/pkg/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/resource/deploy/providers"
+	"github.com/pulumi/pulumi/pkg/resource/plugin"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/pkg/workspace"
+)
+
+const (
+	preparePluginLog        = 7
+	preparePluginVerboseLog = 8
+)
+
+// pluginSet represents a set of plugins.
+type pluginSet struct {
```

Comment:
```suggestion
type pluginSet map[string]workspace.PluginInfo
```

Reply:
Might just want to use a nominative type.

Reply:
will do.